mlcomp/parallelm/mlpiper/wizard_shell.py

```diff
@@ -0,0 +1,383 @@
+            for k, v in ns.__dict__.items():
+                if isinstance(v, six.string_types) and len(v) == 0:
+                    continue
+                dd[k] = v
+
+            component_info.load_from_json(dd)
+
+        else:
+            return EXIT_CODE
+
+        return 0
+
+
+# TODO use args parser for sub commands
+class InputOutputCommand(Command):
+    SUBCOMMAND_DELETE = "delete"
+    SUBCOMMAND_EDIT = "edit"
+    SUBCOMMAND_ADD = "add"
+
+    def __init__(self, name, topic, brief, **kwargs):
+        self.parser = PypsiArgParser(prog=name, description=brief)
+        # TODO: start using args from ArgParser, now they only added to show pretty help
+        self.parser = PypsiArgParser(prog=name, description=brief)
+        subparsers = self.parser.add_subparsers(dest='subparser_name', help="Commands")
+        add_arg_parser = subparsers.add_parser("add", help="add connection")
+        edit_arg_parser = subparsers.add_parser("edit", help="edit connection")
+        edit_arg_parser.add_argument("index", metavar="INDEX", action="store", help="connection index to edit")
+        delete_arg_parser = subparsers.add_parser("delete", help="delete connection")
+        delete_arg_parser.add_argument("index", metavar="INDEX", action="store", help="connection index to delete")
+
+        super(InputOutputCommand, self).__init__(name=name, usage=self.parser.format_help(), topic=topic, brief=brief, **kwargs)
+
+        self._handlers = {self.SUBCOMMAND_ADD: self._add_handler,
+                          self.SUBCOMMAND_DELETE: self._delete_handler,
+                          self.SUBCOMMAND_EDIT: self._edit_handler}
+        self._print_header = True
+
+    def run(self, shell, args):
+        try:
+            ns = self.parser.parse_args(args)
+        except CommandShortCircuit as e:
+            return e.code
+
+        component_info = shell._component_info
+        self._print_header = shell._wizard_edit_mode
+        if self.name == CommandName.INPUT:
+            if component_info._inputs is None:
+                component_info._inputs = []
+            self._inputs_or_outputs_array = component_info._inputs
+        elif self.name == CommandName.OUTPUT:
+            if component_info._outputs is None:
+                component_info._outputs = []
+            self._inputs_or_outputs_array = component_info._outputs
+
+        cmd = self.SUBCOMMAND_ADD
+        if args is not None and len(args):
+            if args[0] not in self._handlers.keys():
+                self.usage_error(shell, args)
+                return 1
+
+            cmd = args[0]
+
+        return self._handlers[cmd](shell, args)
+
+    def _add_handler(self, shell, args=None):
+        comp_conn = ComponentConnectionInfo()
+        w = ConnectionWizard("{} Connection Configuration".format(self.name.capitalize()), comp_conn)
+        ns = w.run(shell, self._print_header)
+        if ns:
+            dd = {}
+            for k, v in ns.__dict__.items():
```

Comment:
A very quick review of the code, I see this kind of code snippet repeats in multiple places. Could you somehow move is to a shared function / parent function?
This comment is relevant to other snippets as well.